src/System.Private.CoreLib/src/System/MulticastDelegate.cs

```diff
@@ -135,32 +136,32 @@ public override sealed bool Equals(object obj)
         private bool InvocationListEquals(MulticastDelegate d)
         {
             Debug.Assert(d != null && (_invocationList as object[]) != null, "bogus delegate in multicast list comparison");
-            object[] invocationList = _invocationList as object[];
+            object[] invocationList = (_invocationList as object[])!; // Calling code ensures invocationList is object[]
```

Comment:
@jcouv, the Debug.Assert above guarantees that `_invocationList as object[]` is not null; is it a bug that the compiler is flagging this (@danmosemsft, I'm assuming it's flagging this and that's why you changed it)?

This predates your change, but I'm also not clear why this is `_invocationList as object[]` rather than `(object[])_invocationList`.  If this cast could fail, it'd null ref below when indexing into it, so it seems it'd be better to just do a normal cast.

Reply:
Right, it cannot infer based on this pattern.

```c#
        // Debug.Assert(_invocationList is Object[]); // does work
        Debug.Assert(_invocationList as Object[] != null);  // does not
        object[] invocationList = (_invocationList as object[]);  // nullability warning
```

https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQDk4BstWFiAEQHsBzMY4gEwBoqQBqAHwGIA7bXfCAAgjbwEAsACgAAgAYeYgIwBuUZOkyAdIQCWUUm2LIY6gMbIF4gMzSATDwDCPAN6ieTngHlgAKwgGYPAPrq2ADdiA1h1YjYAGXU9HgBeHjYIAHceYg8vGABtGQBdE2ceR2cABwR1QNheMQAWHgAxAAoASmKnBxFCwoB6bp5CCGA4UhUAQWRkCAQYRv8gkLCI6NiY1wzvLNzmuR5enipiFB5k4gQAazaugaGR8cnp2YDg0P0lmJ8oZDXPDdyeAEIEhwcNsip0uk50j9sn8ngtXlF3vEeI95i9wojYp80usYdtLgBfUQEoA

Reply:
Excuse me, `Debug.Assert(_invocationList is Object[]);` does not work either. 

https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQDk4BstWFiAEQHsBzMY4gEwBoqQBqAHwGIA7bXfCAAgjbwEAsACgAAgAYeYgIwBuUZOkyAdIQCWUUm2LIY6gMbIF4gMzSATDwDCPAN6ieTngHlgAKwgGYPAPrq2ADdiA1h1YjYAGXU9HgBeHjYIAHceYg8vGABtGQBdE2ceR2cABwR1QNheMQAWHgAxAAoASmKnBxFCwsIIYDhSFQBBZGQIBBhG/yCQsIjo2Ji0jO8s3OaCrp4Aei2eHr6B4dHxyYDg0P05mJ8oZFdl7NyeAEIEjhx1os7Npc8Vp7OM0uUWu8R4p2mF3CINit1+mVW6zaPAAvqIUUA==

Reply:
Logged https://github.com/dotnet/roslyn/issues/34638.

Reply:
> Excuse me, Debug.Assert(_invocationList is Object[]); does not work either.

It seems to work fine here: https://github.com/dotnet/coreclr/pull/23615/files#diff-53dcb7b3c9c67908fe7c008e271cffaaR309 ?

Reply:
The problem seems to be using `as`. If you use it in a `Debug.Assert`, it does not cause the value to be considered non nullable subsequently. Even if you don't use `as`, if you use it in your subsequent expression, it also will still warn.

```c#
#nullable enable
using System;
using System.Diagnostics;
public class C {
    Object? _invocationList = new object[1];
    
    private void F()
    {
        // Debug.Assert(_invocationList as object[] != null);           // does not force non-nullability
        Debug.Assert(_invocationList is object[]);

        //object[] invocationList = (_invocationList as object[]); // nullability warning with either assert
        object[] invocationList = (object[])_invocationList;
    }
}
```